llama_index/chat_engine/types.py

```diff
@@ -195,6 +196,168 @@ async def aprint_response_stream(self) -> None:
+
+    def put_in_queue(self, queue: queue.Queue, delta: Optional[str]) -> None:
+        queue.put_nowait(delta)
+
+    def aput_in_queue(self, queue: asyncio.Queue, delta: Optional[str]) -> None:
+        queue.put_nowait(delta)
+        self._new_item_event.set()
+
+    def write_response_to_history(
+        self, memory: BaseMemory, raise_error: bool = False
+    ) -> None:
+        if self.chat_stream is None:
+            raise ValueError(
+                "chat_stream is None. Cannot write to history without chat_stream."
+            )
+
+        # try/except to prevent hanging on error
+        try:
+            final_text = ""
+            for chat in self.chat_stream:
+                # Cohere response queue
+                self.put_in_queue(self._queue, chat.delta)
+                # Citations queue
+                if chat.raw.get("event_type", "") == StreamEvent.CITATION_GENERATION:
+                    self.put_in_queue(
+                        self._citations_queue, chat.raw.get("citations", [])
+                    )
+                # Documents queue
+                if chat.raw.get("event_type", "") == StreamEvent.SEARCH_RESULTS:
+                    self.put_in_queue(
+                        self._documents_queue, chat.raw.get("documents", [])
+                    )
+                # TODO  Ask TJ - maybe we'll need it
+                final_text += chat.delta or ""
+        except Exception as e:
+            if not raise_error:
+                logger.warning(
+                    f"Encountered exception writing response to history: {e}"
+                )
+            else:
+                raise
+
+        self._is_done = True
+
+        # This act as is_done events for any consumers waiting
+        self._is_function_not_none_thread_event.set()
+
+    async def awrite_response_to_history(
+        self,
+        memory: BaseMemory,
+    ) -> None:
+        if self.achat_stream is None:
+            raise ValueError(
+                "achat_stream is None. Cannot asynchronously write to "
+                "history without achat_stream."
+            )
+
+        # try/except to prevent hanging on error
+        try:
+            final_text = ""
+            async for chat in self.achat_stream:
+                # Cohere response queue
+                self.aput_in_queue(self._aqueue, chat.delta)
+                # Citations queue
+                if chat.raw.get("event_type", "") == StreamEvent.CITATION_GENERATION:
+                    self.aput_in_queue(
+                        self._citations_aqueue, chat.raw.get("citations", [])
+                    )
+                # Documents queue
+                if chat.raw.get("event_type", "") == StreamEvent.SEARCH_RESULTS:
+                    self.aput_in_queue(
+                        self._documents_aqueue, chat.raw.get("documents", [])
+                    )
+                # TODO  Ask TJ - maybe we'll need it
+                final_text += chat.delta or ""
```

Comment:
Same question here

Reply:
As I mentioned in my previous message, we only have a chat.delta for the TEXT_GENERATION event.